python/cudf/cudf/core/udf/utils.py

```diff
@@ -240,6 +240,26 @@ def _construct_signature(frame, return_type, args):
     return sig
 
 
+class Row(Record):
+    # Numba's Record type provides a convenient abstraction for representing a
+    # row, in that it provides a mapping from strings (column / field names) to
+    # types. However, it cannot be used directly since it assumes that all its
+    # fields can be converted to NumPy types by Numba's internal conversion
+    # mechanism (`numba.np_support.as_dtype). This is not the case for cuDF
+    # extension types that might be the column types (e.g. masked types, string
+    # types or group types).
+    #
+    # We use this type for type inference and type checking, but not in code
+    # generation. For this use case, it is sufficient to provide a dtype for a
+    # row that corresponds to any Python object.
+    @property
+    def dtype(self):
+        return np.dtype("object")
```

Comment:
Is it at all useful to try the parent's behavior first and see if it gets a useful dtype, then fall back to object if not?

Reply:
I don't think it's a good idea - presently the dtype we provide is not used, but if it starts being used it wouldn't necessarily be correct to use the parent's dtype anyway - this will break more loudly if it breaks, whereas an issue resulting from using a parent-supplied dtype will look more like weird code-generation bugs.